spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/datastore/GcpDatastoreAutoConfiguration.java

```diff
@@ -102,20 +106,31 @@
 
 	@Bean
 	@ConditionalOnMissingBean
-	public Datastore datastore() {
-		DatastoreOptions.Builder builder = DatastoreOptions.newBuilder()
-				.setProjectId(this.projectId)
-				.setHeaderProvider(new UserAgentHeaderProvider(this.getClass()))
-				.setCredentials(this.credentials);
-		if (this.namespace != null) {
-			builder.setNamespace(this.namespace);
-		}
+	public DatastoreNamespaceProvider namespaceProvider() {
+		return () -> this.namespace;
+	}
 
-		if (this.host != null) {
-			builder.setHost(this.host);
+	@Bean
+	@ConditionalOnMissingBean(value = Datastore.class, parameterizedContainer = Supplier.class)
```

Comment:
Can we make this `@ConditionalOnBean` `namespaceProvider`?

Reply:
If we did that wouldn't that mean there would be an error if the user also tried to define a bean of `Supplier<Datastore>` return type ?

Reply:
Why? Could it still be also conditional on `Supplier<Datastore>` missing?